#Tarea 02

Hacer el código para el método de Gauss-Seidel (o Jacobi) y dar una aproximación a la solución para el siguiente sistema de ecuaciones:


\begin{align*}
x_4 + 3x_5 &= 3 \\
2x_1 + x_2 - x_3 &= 5 \\
4x_3 + 2x_4 - 6x_5 &= 10 \\
3x_1 + 5x_2 + 2x_3 - x_4 &= 12 \\
x_2 - 3x_3 + 4x_4 - x_5 &= -8 \\
\end{align*}

\

Aplicar una tolerancia en el error aproximado relativo porcentual de 5%

In [ ]:
#MÉTODO DE GAUSS-SEIDEL

import numpy as np

def Gauss_Seidel(a, b, iteraciones, tol):

  n = len(b)
  x = np.zeros(n)
  soluciones = []

  k = 1

  while k <= iteraciones:

    x_0 = x.copy()

    for i in range(n):
      suma = 0
      for j in range(n):
        if j != i:
          suma += a[i,j]*x[j]
      x[i] = (1/a[i,i])*(b[i] - suma)

    print("Iteración:", k)
    print("Soluciones =", x)
    print('\n')

    error_relativo_porcentual = (np.abs(x - x_0)/np.abs(x))*100

    if np.all(error_relativo_porcentual < tol):
      print(f'El metodo converge despues de {k} iteraciones para la tolerancia dada.\nLa solucion es:', x)
      soluciones.extend(x)
      return soluciones

    else:
      k += 1

  print(f'El metodo fallo despues de {k-1} iteraciones')

In [ ]:
#MÉTODO EMPLEADO PARA EJEMPLO DEL LIBRO (Burden, página 338, ejemplo 3)

M = np.array(([10, -1, 2, 0], [-1, 11, -1, 3], [2, -1, 10, -1], [0, 3, -1, 8]), dtype=float)
N = np.array(([6, 25, -11, 15]), dtype=float)

Gauss_Seidel(M, N, 100, 0.05)

Iteración: 1
Soluciones = [ 0.6         2.32727273 -0.98727273  0.87886364]


Iteración: 2
Soluciones = [ 1.03018182  2.03693802 -1.0144562   0.98434122]


Iteración: 3
Soluciones = [ 1.00658504  2.00355502 -1.00252738  0.99835095]


Iteración: 4
Soluciones = [ 1.00086098  2.00029825 -1.00030728  0.99984975]


Iteración: 5
Soluciones = [ 1.00009128  2.00002134 -1.00003115  0.9999881 ]


Iteración: 6
Soluciones = [ 1.00000836  2.00000117 -1.00000275  0.99999922]


El metodo converge despues de 6 iteraciones para la tolerancia dada.
La solucion es: [ 1.00000836  2.00000117 -1.00000275  0.99999922]


[1.0000083636613348, 2.000001173336268, -1.0000027450726756, 0.999999216864815]

In [ ]:
#MÉTODO EMPLEADO PARA SISTEMA DE ECUACIONES PROPUESTO

A = np.array(([2, 1, -1, 0, 0], [3, 5, 2, -1, 0], [0, 1, -3, 4, -1], [0, 0, 4, 2, -6], [0, 0, 0, 1, 3]), dtype=float)
B = np.array(([5, 12, -8, 10, 3]), dtype=float)

Gauss_Seidel(A, B, 100, 0.05)

Iteración: 1
Soluciones = [ 2.5         0.9         2.96666667 -0.93333333  1.31111111]


Iteración: 2
Soluciones = [ 3.53333333 -1.09333333  0.62074074  7.69185185 -1.56395062]


Iteración: 3
Soluciones = [  3.35703704   1.67585185  14.00240329 -27.69665844  10.23221948]


Iteración: 4
Soluciones = [  8.66327572 -13.93825844 -42.31903722 120.33473288 -39.11157763]


Iteración: 5
Soluciones = [ -11.69038939   50.4087951   192.95310141 -498.24093569  167.0803119 ]


Iteración: 6
Soluciones = [  73.77215316 -218.6927196  -790.24559142 2086.73211854 -694.57737285]


Iteración: 7
Soluciones = [ -283.27643591   905.81052183  3318.43878962 -8715.60969778
  2906.20323259]


Iteración: 8
Soluciones = [  1208.81413389  -3793.38593574 -13851.34265315  36426.29500407
 -12141.09833469]


Iteración: 9
Soluciones = [  -5026.4783587    15844.08307729   57899.45380942 -152217.20262291
   50740.06754097]


Iteración: 10
Soluciones = [  21030.18536606  -66218.93326799 -241939.93710019  636105.07682329
 

Se observa que el método de Gauss-Seidel no converge por sí solo después de muchas iteraciones; por lo tanto, este método falla para este sistema. Podemos buscar como solución reacondicionar el sistema de ecuaciones para hacerlo más susceptible a la convergencia. En este caso, factorizaremos la matriz de coeficientes $A$ mediante la factorización $LU$ y resolveremos el sistema haciendo uso de esta factorización y el método de Gauss-Seidel.

Mediante la factorización $LU$, podemos resolver el sistema de la siguiente forma. Sea entonces:

$A$**x** $=$ **b**, donde $A = LU$, entonces:

$A$**x** $=$ $LU$**x** $=$ **b**

Introducimos la sustitución **y** $=$ $U$**x**. Entonces **b** $=$ $L$($U$**x**) $=$ $L$**y**. Es decir:

$L$**y** $=$ **b**

De este modo, obtenemos dos sistemas de ecuaciones que requerimos resolver, uno para **y**, y otro para **x**:

$L$**y** $=$ **b**

$U$**x** $=$ **y**

Entonces, una vez factorizada la matriz $A$ de coeficientes en la forma $LU$, resolveremos los sistemas de ecuaciones anteriores mediante el método de Gauss-Seidel para obtener una aproximación que converja correctamente.

In [ ]:
#FACTORIZACIÓN LU

import numpy as np

def LU(a):
  n = len(a)
  L = np.zeros((n, n))
  U = np.zeros((n, n))

  for i in range(n):
    L[i, i] = 1

    for j in range(i, n):
      U[i, j] = a[i, j]
      for k in range(i):
        U[i, j] -= L[i, k] * U[k, j]

    for j in range(i+1, n):
      L[j, i] = a[j, i]
      for k in range(i):
        L[j, i] -= L[j, k] * U[k, i]
      L[j, i] /= U[i, i]

  return L, U

In [ ]:
#Factorización de 'A' en LU
A = np.array(([2, 1, -1, 0, 0], [3, 5, 2, -1, 0], [0, 1, -3, 4, -1], [0, 0, 4, 2, -6], [0, 0, 0, 1, 3]), dtype=float)
B = np.array(([5, 12, -8, 10, 3]), dtype=float)

L, U = LU(A)

In [ ]:
y = Gauss_Seidel(L, B, 100, 0.05) #Solución del sistema 'Ly = b'

Iteración: 1
Soluciones = [ 5.          4.5        -9.28571429  0.71428571  2.88636364]


Iteración: 2
Soluciones = [ 5.          4.5        -9.28571429  0.71428571  2.88636364]


El metodo converge despues de 2 iteraciones para la tolerancia dada.
La solucion es: [ 5.          4.5        -9.28571429  0.71428571  2.88636364]


In [ ]:
x = Gauss_Seidel(U, y, 100, 0.05) #Solución del sistema 'Ux = y' (obtención de las soluciones finales)

Iteración: 1
Soluciones = [2.5        1.28571429 2.32142857 0.11363636 0.70165746]


Iteración: 2
Soluciones = [ 3.01785714 -1.00324675  2.26776745  0.89502762  0.70165746]


Iteración: 3
Soluciones = [ 4.1355071  -0.72633099  3.10497238  0.89502762  0.70165746]


Iteración: 4
Soluciones = [ 4.41565168 -1.56353591  3.10497238  0.89502762  0.70165746]


Iteración: 5
Soluciones = [ 4.83425414 -1.56353591  3.10497238  0.89502762  0.70165746]


Iteración: 6
Soluciones = [ 4.83425414 -1.56353591  3.10497238  0.89502762  0.70165746]


El metodo converge despues de 6 iteraciones para la tolerancia dada.
La solucion es: [ 4.83425414 -1.56353591  3.10497238  0.89502762  0.70165746]


Dadas las soluciones, comprobamos los resultados haciendo uso de la función `np.linalg.solve` de NumPy, considerando esos resultados como correctos.

In [ ]:
import numpy as np

A = np.array(([2, 1, -1, 0, 0], [3, 5, 2, -1, 0], [0, 1, -3, 4, -1], [0, 0, 4, 2, -6], [0, 0, 0, 1, 3]), dtype=float)
B = np.array(([5, 12, -8, 10, 3]), dtype=float)

x = np.linalg.solve(A, B)
print(x)

[ 4.83425414 -1.56353591  3.10497238  0.89502762  0.70165746]


Finalmente, después de haber reacondicionado el sistema, pudimos lograr que el método convergiera correctamente. Después de verificar los resultados, encontramos las soluciones del sistema de ecuaciones:

*   $x_1 = 4.8342$
*   $x_2 = -1.5635$
*   $x_3 = 3.1049$
*   $x_4 = 0.8950$
*   $x_5 = 0.7016$

